### DISCUSSION TITLE
Testable `assert` function

### DISCUSSION THREAD

**Discussion by john-flanagan** [tgrapperon/swift-dependencies-additions#17] [Ideas]
I opened [this PR](https://github.com/pointfreeco/swift-dependencies/pull/7) on the `swift-dependencies` library about making a testable version of `assert` using `DependencyValues`, but based on the discussion it sounds like it'd be a better fit here.

I just wanted to kick off a discussion about adding it to this library.

**Comment by tgrapperon** [comment]
Hey @john-flanagan! Thanks for opening the discussion!

When I first saw your PR, I thought about extending its capabilities to `precondition` or even `fatalError`, but after thinking more about it, I realized that it could lead to security issues if someone ever manages to override or delay the crashing behavior in some way or another. 

`assert` is less concerning because it is stripped out of `Release` builds, but I have a few open questions:
- Do you think it's worth making this dependency configurable? I feel that its main interest is to degrade very nicely when testing. Baking this behavior once and for all would probably make it a little "safer", without impeding too much its general attractiveness.
- Do you think we can expand it to also generate [assertion failures](https://developer.apple.com/documentation/swift/assertionfailure(_:file:line:)), and offer a way to act more imperatively if required? For example with a `func failure(…)`. In this case, should we base this implementation on the predicate's one, or the other way (which is where I'd be leaning into)?
- Should we `noop` (through `testValue`) for the `previewValue`? Crashes in SwiftUI previews are pretty disruptive, but I don't know if falling through silently is the correct approach, especially since we don't have a console to check what's going on.

For future reference, it seems we would be able to dynamically find the current `XCTestCase` context and switch `continueAfterFailure = false` when failing, so the test method wouldn't continue after the "crash". But this would be mostly internal and can be improved in a later PR.

Let me know what you think!

**Reply by KaiOelfke** [reply]
I have a debug scenario where I can't use OSLog as I need logs from multiple app launches. I use some other logging tool, which is writing async with GCD to a file. And the feature involves a background URLSession so Xcode / a debugger can't be attached. Normal asserts functions are problematic here as they crash the app without a log. So it'd be nice to temporarily override the assert here to just log to my logging framework.

This is just to share a theoretical use case. I can't use the Dependency system anyway in this project due to unrelated reasons.

Sometimes I also commented out assertions temporarily in bigger projects, if it's a known issue being addressed elsewhere.

A unit test may wants to verify that something asserts a thing. E.g. I made one property wrapper @PreMainActor for code I can't yet update to Concurrency, but that should be `@MainActor` in the future. When running on a non main thread this should XCTFail in a test and assert during debug runs. I have tests for initializing, getting and setting properties wrapped with this.

**Reply by tgrapperon** [reply]
Yeah, that's a great example. The fact that `assert` behavior changes in release also confirms that crashing is not the only intended outcome, it it should probably be made configurable like a standard dependency. It should also probably noop in previews.
@john-flanagan Do you want to PR this feature? It can be your previous implementation only, and I'll make the few adjustments to handle assertion failures if you prefer.

**Comment by john-flanagan** [comment]
@tgrapperon, I just opened [this PR](https://github.com/tgrapperon/swift-dependencies-additions/pull/28) to add a basic `assert` and `assertionFailure` dependency!
